### ISSUE TITLE
DL4J extensibility

### ISSUE THREAD

**Issue by EronWright** [deeplearning4j/deeplearning4j#251]
Putting in place a foundation for extensibility is definitely worthwhile, given the many conceivable extension points (layers, loss functions, preprocessors, step functions, solvers, distributions, etc.).

I propose we follow the Jackson approach of 'modules' ([more info](http://wiki.fasterxml.com/JacksonFeatureModules)).  

**Registration:** A module shall be registered either explicitly or via auto-discovery (ServiceLoader).   The API shall consist of 'findModules' (to scan the classpath) and 'registerModule' (to add a module), with a convenience method of 'findAndRegisterModules'. 

**Lifecycle:**  Modules must be registered before configurations are loaded, saved, or executed.   Today, DL4J has no top-level context object to which a module may be registered (analogous to ObjectMapper).   *To be decided is whether to introduce a context object or to rely on static registration.*

**SPI:**  During registration, the module shall be provided a 'SetupContext', allowing for registration of layer types, loss functions, etc.   See [example](http://fasterxml.github.io/jackson-databind/javadoc/2.3.0/com/fasterxml/jackson/databind/Module.SetupContext.html).  The SetupContext interface will naturally evolve with new extension points over time.

**Configurability:**  Extensions shall participate in both the configuration of neural networks and subsequent execution.   It shall be possible to register custom configuration elements (such as those found in o.d.nn.conf.*) to be handled in a first-class manner in serialized/JSON configurations.   The underlying Jackson ObjectMapper shall be programmed accordingly.

**Comment by agibsonccc** [comment]
We will revisit this with the computation graph. I think focusing on extensions via that api will be better.
